src/EventListener/OrderPaymentPaidListener.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\InvoicingPlugin\EventListener;
+
+use Knp\Snappy\GeneratorInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Sylius\InvoicingPlugin\Email\InvoiceEmailSenderInterface;
+use Sylius\InvoicingPlugin\Entity\InvoiceInterface;
+use Sylius\InvoicingPlugin\Event\OrderPaymentPaid;
+use Sylius\InvoicingPlugin\Repository\InvoiceRepository;
+use Symfony\Component\Templating\EngineInterface;
+
+final class OrderPaymentPaidListener
+{
+    /** @var InvoiceRepository */
+    private $invoiceRepository;
+
+    /** @var OrderRepositoryInterface */
+    private $orderRepository;
+
+    /** @var GeneratorInterface */
+    private $pdfGenerator;
+
+    /** @var InvoiceEmailSenderInterface */
+    private $emailSender;
+
+    /** @var EngineInterface */
+    private $templatingEngine;
+
+    public function __construct(
+        InvoiceRepository $invoiceRepository,
+        OrderRepositoryInterface $orderRepository,
+        GeneratorInterface $pdfGenerator,
+        InvoiceEmailSenderInterface $emailSender,
+        EngineInterface $templatingEngine
+    ) {
+        $this->invoiceRepository = $invoiceRepository;
+        $this->pdfGenerator = $pdfGenerator;
+        $this->emailSender = $emailSender;
+        $this->orderRepository = $orderRepository;
+        $this->templatingEngine = $templatingEngine;
+    }
+
+    public function __invoke(OrderPaymentPaid $event)
+    {
+        /** @var InvoiceInterface $invoice */
+        $invoice = $this->invoiceRepository->findOneByOrderNumber($event->orderNumber());
+
+        $filePath = $this->generateTemporaryPdfFilePathBasedOnInvoiceId($invoice->id());
+
+        $this->pdfGenerator->generateFromHtml(
+            $this->templatingEngine->render('@SyliusInvoicingPlugin/Resources/views/Invoice/Download/pdf.html.twig', [
+                'invoice' => $invoice
+            ]),
+            $filePath
```

Comment:
It looks kinda weird right now to have this file created implicitly (`$this->pdfGenerator->generateFromHtml(...)`) and then removing it explicitly (`$this->removeTemporaryPdfFile(...);`).

My suggestion would be to extract file-related operations from this class, making it something like:

```php
$invoiceContent = $this->pdfGenerator->generateOutputFromHtml(...);

$this->temporaryFilesystem->create($file, $invoiceContent);

try {
  $this->emailSender->sendInvoiceEmail();
} finally {
  $this->temporaryFilesystem->remove($file);
}
```

Usage of that `finally` block will make sure that even if sending that email fails, the invoice gets removed afterwards.

As for the future, I'd think about extracting this whole listener logic to a command handler (and just dispatching SendInvoiceEmail command), so that it will be possible to handle it asynchronously and the logic will be separated (listening on event / reacting on event).

Reply:
Btw. this class is missing a unit test :)